### PR TITLE
Update examples (+ bug fix)

### PR DESCRIPTION
**Pull request by lcarver**
Hello,
After the recent big merges, I wanted to update the example files to include the beam monitor, and also take into account all the new multibunch features which were not all implemented properly. 

During this update, I found a few extra things. One was in wake_object, the shunt impedance was passed as the frequency which was causing a nonsensical wake function. Easy fix.
I also implemented directly in the factory function for the longitudinal resonator, a point forced at 1e-24. This is to avoid interpolation issues around 0, which is only a problem for the longitudinal wake as it has the fundamental beam loading theorem. This had to be considered in the haissinski file too.

### CODE REVIEW

**Comment by lcarver** [comment]
So, it can be cleaned up yes but build_srange is not the place to do it in my opinion.
The problem arises from the specific case of the longitudinal resonator function which has discontinuities:

for 
```
s<0 W=0
s=0 W=a/2
s>0 W = a * func(t)
```
In the wakefieldpass passmethod we use interpolation between the points provided by srange. Each dirty line that you mentioned solves the problem on either side of the discontinuity. The clip sets all negative s to 0, then unique removes them. The addition of 1e-24 to s ensures that the interpolation does not provided incorrect values for small s. 

Note that for the beam loading pass method the wake function is implemented directly in C without interpolation so it is not a problem for that case.

**Comment by lfarv** [comment]
Ok, but `unique(clip())` is the most inefficient way of getting rid of negative values ! It implies sorting the full array… And I don't see why `build_srange` could not produce ranges without negative values ? Am I missing something ?

Anyway, this is not related to the update of examples, which is a good point !

**Comment by lfarv** [comment]
> May be this is inefficient but this is called only once during initialization so the penalty is negligible.

I was simply noting that simply looping through srange up to the 1st positive value would be much faster that sorting the whole array ! You know it'a already sorted.

> Another option is to raise an error if the user tries to model a wake function using negative values and 0.0 is not present in the table. This is certainly educative as it forces the user to understand what he is doing but is as complicated as forcing it in my opinion...

For me this is by far the best solution. The doc for longitudinal wakes should state that **srange must start at zero**, and it looks simple to check the 1st value in srange. It's better than silently altering the user's choice.

Same for the magic 1e-24: if it's crucial, why not systematically putting it in srange?

**Comment by swhite2401** [comment]
> Same for the magic 1e-24: if it's crucial, why not systematically putting it in srange?

We could, but it would still a problem in case the user decides to build srange using `numpy.linspace()` for instance... by the way @lcarver the way it is added could be problematic as `srange` should have unique values.

**Comment by swhite2401** [comment]
@lcarver, I see the following code in the RW wake function:

```
    if numpy.amin(srange) <= 0:
        raise ValueError("""
                        Provided srange has either negative values or 0s
                        This is not allowed for the transverse resistive wall
                        wake function. Please correct.
                        """)
```

Can we add it for all wake functions (resonator should start at zero)

I would also just use `srange[0] <=0` and put as error "srange should be ordered and containing only unique positive values" or something like this. To be adapted for the resonator

**Comment by swhite2401** [comment]
@lcarver, if you allow me I would like to fix #533 in the PR as well

**Comment by carmignani** [comment]
Fixing the mexFunctions is easy, I can also do if you want, but it would be good also to fix the automatic GitHub tests. atmexall is called, it fails in the automatic check, but still the test result is green.

**Comment by lfarv** [comment]
> atmexall is called, it fails in the automatic check, but still the test result is green.

In fact, `atmexall` does not fail: it's like on your computer, if one MexFunction fails to compile, `atmexall` ignores it and goes on. A real test would be to try and execute the passmethod in `pytests.m`.

For general use, failing on the 1st wrong MexFunction is unconvenient. But I can look for adding a flag to force this behaviour in GitHub tests.

**Comment by swhite2401** [comment]
> Fixing the mexFunctions is easy, I can also do if you want

It is done, should be fine now (on this branch)

**Comment by swhite2401** [comment]
@lcarver, are working on this one? Do you want me to implement the requested changes?

**Comment by lcarver** [comment]
```
import at
import numpy as np

ring = at.load_lattice('../../../machine_data/esrf.m')
srange0 = np.linspace(-0.1, 0.1, 11)
srange1 = np.linspace(0.1, 0.2, 11)

```
Then: 
`p0 = at.LongResonatorElement('a', ring, srange0, 1, 1, 1)
`
gives an error.

`p1 = at.LongResonatorElement('a', ring, srange1, 1, 1, 1)
print(p1._srange)
`
gives an array which contains 0 and 1e-24 at the beginning. 

The same tests were made for transverse element and for reswall, and it is working fine.
I removed the warnings from the wake_functions themselves because if the functions are called directly then it is at the risk of the user.

File: atintegrators/BeamLoadingCavityPass.c [hunk 1]
```diff
@@ -270,9 +276,9 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
       double *bcurr = malloc(sizeof(double));
       bspos[0] = 0.0;
       bcurr[0] = 0.0;
-      BeamLoadingCavityPass(r_in,num_particles,1,bspos,bcur,1,0,Elem);
+      BeamLoadingCavityPass(r_in,num_particles,1,bspos,bcurr,1,0,Elem);
       free(bspos);
-      free(bcur);
+      free(bcurr);
```
[the remark this answers is not in the document]
Comment (lcarver) [reply atintegrators/BeamLoadingCavityPass.c]: done

**Comment by swhite2401** [comment]
@lcarver, to be safe you need to check that `srange` is unique in the `__init__` function of `wake_object`

**Comment by lcarver** [comment]
> @lcarver, to be safe you need to check that `srange` is unique in the `__init__` function of `wake_object`

done

**Comment by lcarver** [comment]
assertion added so that user is forced to provide unique values. The checks were also removed as no longer needed.

**Comment by swhite2401** [comment]
Ok after the last pending item has been fixed (simplified addition of 0.0 and 1e-24 -> see code review)

**Comment by swhite2401** [comment]
Ah one last thing, could you please update the format of the docstring of `haissinski.py` such that it coherent with other classes (definition of argument, typing, etc...)

**Comment by lcarver** [comment]
Resolved all except the Haissinski which I will do tomorrow as I am off this afternoon.

**Comment by swhite2401** [comment]
@lcarver, I have corrected a couple other bugs that I found

**Review by swhite2401 — APPROVED** [review]
Ok !

**Pull request merged**